src/structure_managers/property_block_sparse.hh

```diff
@@ -0,0 +1,600 @@
+      using MyMap_t = std::map<K, V>;
+
+      //! the data holder. only the overiden/essential functionalities are
+      //! directly exposed
+      MyMap_t data{};
+
+      // some member types
+      using key_type = typename MyMap_t::key_type;
+      using mapped_type = typename MyMap_t::mapped_type;
+      using value_type = typename MyMap_t::value_type;
+      using size_type = typename MyMap_t::size_type;
+      using reference = typename MyMap_t::reference;
+      using const_reference = typename MyMap_t::const_reference;
+      using iterator = typename MyMap_t::iterator;
+      using const_iterator = typename MyMap_t::const_iterator;
+
+      //! Default constructor
+      InternallySortedKeyMap() = default;
+
+      //! Copy constructor
+      InternallySortedKeyMap(const InternallySortedKeyMap & other) = default;
+
+      //! Move constructor
+      InternallySortedKeyMap(InternallySortedKeyMap && other) = default;
+
+      //! Destructor
+      ~InternallySortedKeyMap() = default;
+
+      //! Copy assignment operator
+      InternallySortedKeyMap &
+      operator=(const InternallySortedKeyMap & other) = default;
+
+      //! Move assignment operator
+      InternallySortedKeyMap &
+      operator=(InternallySortedKeyMap && other) = default;
+
+      /**
+       * Returns a reference to the mapped value of the element with key
+       * equivalent to key. If no such element exists, an exception of type
+       * std::out_of_range is thrown.
+       * The elements of the key are sorted in ascending order.
+       *
+       */
+      mapped_type & at(const key_type & key) {
+        key_type skey{this->copy_sort(key)};
+        return this->data.at(skey);
+      }
+      const mapped_type & at(const key_type & key) const {
+        key_type skey{this->copy_sort(key)};
+        return this->data.at(skey);
+      }
+      //! access or insert specified element
+      mapped_type & operator[](const key_type & key) {
+        key_type skey{this->copy_sort(key)};
+        return this->data[skey];
+      }
+      mapped_type & operator[](key_type && key) {
+        key_type skey{this->copy_sort(key)};
+        return this->data[skey];
+      }
+
+      //! Returns the number of elements with key that compares equivalent to
+      //! the specified argument, which is either 1 or 0 since this container
+      //! does not allow duplicates.
+      template <class Key>
+      decltype(auto) count(const Key & key) {
+        key_type skey{this->copy_sort(key)};
+        return this->data.count(key);
+      }
+
+      //! Erases all elements from the container. After this call, size()
+      //! returns zero.
+      void clear() noexcept { this->data.clear(); }
+
+      template <typename... Args>
+      decltype(auto) emplace(Args &&... args) {
+        return this->data.emplace(std::forward<Args>(args)...);
+      }
+
+      /**
+       * returns a vector of the valid keys of the map
+       */
+      std::vector<key_type> get_keys() {
```

Comment:
I only stumbled over it here - What can `key_type` all mean? It is invoked with a class template `K` - is this intentional?